h/views/api_auth.py

```diff
@@ -2,11 +2,100 @@
+
+    @view_config(request_method='GET',
+                 renderer='h:templates/oauth/authorize.html.jinja2')
+    def get(self):
+        """
+        Check the user's authentication status and present the authorization
+        page.
+        """
+        self._check_params()
+
+        if self.request.authenticated_userid is None:
+            raise HTTPFound(self.request.route_url('login', _query={
+                              'next': self.request.url}))
+
+        params = self.request.params
+        user = self.user_svc.fetch(self.request.authenticated_userid)
+
+        return {'username': user.username,
+                'client_name': 'Hypothesis',
+                'client_id': params['client_id'],
+                'response_type': params['response_type'],
+                'response_mode': params['response_mode'],
+                'state': params.get('state')}
+
+    @view_config(request_method='POST',
+                 renderer='h:templates/oauth/post_authorize.html.jinja2',
+                 effective_principals=security.Authenticated)
+    def post(self):
+        """
+        Process an authentication request and return an auth code to the client.
+
+        Depending on the "response_mode" parameter the auth code will be
+        delivered either via a redirect or via a `postMessage` call to the
+        opening window.
+        """
+        authclient = self._check_params()
+
+        user = self.user_svc.fetch(self.request.authenticated_userid)
+
+        # Create an "authorization code" for the response.
+        # This is in fact just a JWT grant token since auth codes have not yet
+        # been implemented.
+        auth_code = self.oauth_svc.create_grant_token(user, authclient)
+
+        params = self.request.params
+
+        return {'code': auth_code,
+                # Once authclients have an Origin property, that should be used
+                # instead here.
+                'origin': self.request.host_url,
+                'state': params.get('state')}
+
+    def _check_params(self):
+        """
+        Check parameters for the authorization request.
+
+        If the parameters are valid, returns an authclient.
+        Otherwise, raises an exception.
+        """
+        params = self.request.params
+
+        client_id = params.get('client_id', '')
+        authclient = self.oauth_svc._get_authclient_by_id(client_id)
+        if not authclient:
+            raise HTTPBadRequest('Unknown client ID "{}"'.format(client_id))
+
+        response_type = params.get('response_type')
+        if response_type != 'code':
+            raise HTTPBadRequest('Unsupported response type "{}"'
+                                 .format(response_type))
+
+        response_mode = params.get('response_mode', 'query')
+        if response_mode != 'web_message':
+            raise HTTPBadRequest('Unsupported response mode "{}"'.format(response_mode))
+
+        return authclient
```

Comment:
I wondered if `_check_params()` should just return the checked params (`response_mode` and `response_type`) because the methods that call `_check_params()` then end up having to get these out of the `request` again anyway